### DISCUSSION TITLE
Is it possible to use this outside of Kotlin, in particular I want to use this in a Scala app (but a Java example also suffices)

### DISCUSSION THREAD

**Discussion by DamianReeves** [cashapp/zipline#1321] [Q&A]
Are there examples of how to use this outside of Android on the JVM? I'm interested in seeing if I can use this in my Scala CLI and server app.

**Comment by JakeWharton** [comment]
You can use the JS engine directly through the `QuickJs` type on any JVM language, but you'll be dealing with APIs that basically only deal in strings. In order to use the type-safe binding of interfaces on the host to implementations in the guest (what we call the JS), only Kotlin is supported as the binding is created using a Kotlin compiler plugin to rewrite the callsites.

**Reply by DamianReeves** [reply]
Do you mind pointing me to the Kotlin code in question, I may be able to use some Scala macro magic to simulate the idea.
Also are there any writeups or blogposts I can look at for examples?

**Reply by JakeWharton** [reply]
The Kotlin plugin itself is here: https://github.com/cashapp/zipline/tree/trunk/zipline-kotlin-plugin. Many of its internals will feature comments showing the shape of code that it is generating. It is very verbose internally because Kotlin IR is already verbose, but we're also generating some fairly complex things syntactically that require a lot of IR.

There are also multiple examples of what the class looks like after manipulation here: https://github.com/cashapp/zipline/blob/6f4a230ee9a9348624e732599e173d50e76fce8e/zipline-testing/src/hostMain/kotlin/app/cash/zipline/testing/ZiplineTestInternals.kt#L97-L224.

Another important thing to note is that the library relies on kotlinx.serialization which is itself a Kotlin compiler plugin that transforms code. So even if you mimic our transformation, there may still be transformations around serialization that are needed.
